src/crypto/openssl/openssl_accel_engine.hpp

```diff
@@ -0,0 +1,84 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (c) 2020 Huawei Technologies Co., Ltd.
+ *
+ * Author: Hang Li <lihang48@huawei.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ */
+
+#ifndef OPENSSL_ACCEL_ENGINE_H
+#define OPENSSL_ACCEL_ENGINE_H
+
+#include "common/debug.h"
+#include "global/global_context.h"
+
+#include <openssl/engine.h>
+
+// -----------------------------------------------------------------------------
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_crypto
+#undef dout_prefix
+#define dout_prefix engine_prefix(_dout)
+
+static ostream& engine_prefix(std::ostream* _dout) {
+  return *_dout << "OpensslAccelEngine: ";
+}
+// -----------------------------------------------------------------------------
+
+class OpensslAccelEngine {
+ private:
+  explicit OpensslAccelEngine(string eid) {
+    if (eid != "none") {
+      ENGINE_load_dynamic();
+
+      std::unique_ptr<ENGINE, std::function<void(ENGINE*)>> engine = {
+          ENGINE_by_id(eid.c_str()), [](ENGINE* e) { ENGINE_free(e); }};
+
+      if (engine.get()) {
```

Comment:
this code is convoluted. better to refactor it like
```c++

ENGINE* load_engine(const char* id)
{
  ENGINE *e = ENGINE_by_id(eid.c_str());
  if (!e) {
    return nullptr;
  }
  if (!ENGINE_init(e)) {
    ENGINE_free(e)
    return nullptr;
  }
  return e;
}

void free_engine(ENGINE* e)
{
  if (e) {
    ENGINE_finish(e);
    ENGINE_free(e);
  }
}

using engine_ptr = std::unique_ptr<ENGINE, std::function<void(ENGINE*)>>;

engine_ptr make_engine(const string& id)
{
  return {load_engine(id.c_str()), free_engine};
}
```
and only return a non-null `CryptoAccelRef` if engine can be created and initialised. otherwise we have no way to pass the failure up to `AES_256_CBC::cbc_transform()`.

Reply:
@tchaikov  this looks good,  but i wonder can we still make every type of engine a sigleton if moving `ENGINE_by_id` outside the constructor,  because operations like MD5 and AES may use hardware engines in different ways.   
once multiple instances of engine are created by same id (which usually share one memory address) in one process,  when one is freed, others may be influenced. that's what i am worried about.

Reply:
@egggHang sure. i am not trying to enable us to have multiple OpenSSL engines or crypto plugins, i just want to break this code chunk into smaller pieces in hope to improve the readability.